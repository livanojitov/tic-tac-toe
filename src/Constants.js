const EMPTY = 0;
const COMPUTER = 1;
const USER = 2;
const TOTAL_SQUARES = 9;
const RANDOM = 1;
const HARD = 2;
const LOST = 0;
const WON = 1;
const DRAW = 2;
const UNKNOWN = 3;
const ENGLISH = 0;
const FRENCH = 1;
const SPANISH = 2;
const IMAGES_PER_CATEGORY = 20;

export {
  EMPTY,
  COMPUTER,
  USER,
  RANDOM,
  HARD,
  TOTAL_SQUARES,
  LOST,
  WON,
  DRAW,
  UNKNOWN,
  ENGLISH,
  FRENCH,
  SPANISH,
  IMAGES_PER_CATEGORY,
};
