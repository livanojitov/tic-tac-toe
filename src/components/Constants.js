const EMPTY = 0;

const COMPUTER = 1;
const USER = 2;

const EASY = 1;
const HARD = 2;
const NORMAL = 3;

const TOTAL_SQUARES = 9;

const CATEGORY = 'Category';

const YOU = 'You';
const OPPONENT = 'Computer';
const REFRESH = 'Refresh';
const USER_IMAGE_TOOLTIP = 'You will play with this image';
const COMPUTER_IMAGE_TOOLTIP = 'The computer will play with this image';
const REFRESH_TOOLTIP = 'Click to refresh the two images';

const START = 'Start';
const YES = 'Yes';
const NO = 'No';
const START_TOOLTIP = 'Who starts playing the game: You or the computer?';
const YES_TOOLTIP = 'You start playing the game.';
const NO_TOOLTIP = 'The computer starts playing the game.';
const GAME_STARTED_BY = 'Game started by:';

const LEVEL = 'Level';
const COMPUTER_PLAYS_EASY = 'Easy';
const COMPUTER_PLAYS_HARD = 'Hard';
const COMPUTER_PLAYS_NORMAL = 'Normal';
const LEVEL_TOOLTIP = 'How much thinking the computer does before playing.';
const COMPUTER_PLAYS_EASY_TOOLTIP = 'The computer plays randomly.';
const COMPUTER_PLAYS_HARD_TOOLTIP = 'The computer does some thinking before playing.';
const COMPUTER_PLAYS_NORMAL_TOOLTIP = 'The computer sometimes plays easy and sometimes plays hard.';
const LEVEL_HISTORY_EASY = 'Level: Easy';
const LEVEL_HISTORY_HARD = 'Level: Hard';
const LEVEL_HISTORY_NORMAL = 'Level: Normal';

const PLAY = 'Start the game';
const GAME_OVER = 'Game over';
const PLAY_AGAIN = 'Play again';
const LOST = 'You lost!';
const WON = 'You won!';
const DRAW = 'It\'s a draw!';

const GAME_WRITTEN_WITH = 'This game is written with:';
const SOURCE_CODE = 'Source Code:';
const OTHER_GAMES = 'Other games:';

const AUTHOR = 'Author:';
const EMAIL = 'Email:';
const GITHUB = 'Github:';
const LINKEDIN = 'LinkedIn:';
const HIRE_ME = 'Hire me!';

const NO_HISTORY_YET = 'No history yet. Play some games and comeback.';

const HOME = 'Home';
const HISTORY = 'History';
const ABOUT = 'About';
const CONTACT = 'Contact';

export { EMPTY, COMPUTER, USER, EASY, HARD, NORMAL, TOTAL_SQUARES, CATEGORY, YOU, OPPONENT, REFRESH, USER_IMAGE_TOOLTIP, COMPUTER_IMAGE_TOOLTIP, REFRESH_TOOLTIP, START, YES, NO, START_TOOLTIP, YES_TOOLTIP, NO_TOOLTIP, GAME_STARTED_BY, LEVEL, COMPUTER_PLAYS_EASY, COMPUTER_PLAYS_HARD, LEVEL_TOOLTIP, COMPUTER_PLAYS_EASY_TOOLTIP, COMPUTER_PLAYS_HARD_TOOLTIP, LEVEL_HISTORY_EASY, LEVEL_HISTORY_HARD, PLAY, GAME_OVER, PLAY_AGAIN, LOST, WON, DRAW, GAME_WRITTEN_WITH, SOURCE_CODE, OTHER_GAMES, AUTHOR, EMAIL, GITHUB, LINKEDIN, HIRE_ME, NO_HISTORY_YET, HOME, HISTORY, ABOUT, CONTACT, COMPUTER_PLAYS_NORMAL, COMPUTER_PLAYS_NORMAL_TOOLTIP, LEVEL_HISTORY_NORMAL };  
