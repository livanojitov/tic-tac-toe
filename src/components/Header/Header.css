.header{
  width: 95%;
  margin: 0 auto
}

header{ 
  display:inline-block;
  width: 30%;
}

.flags img {
  cursor: pointer;
  margin: 0 2%;
}

@media screen and (max-width: 400px) {
  header{
    vertical-align: top;
    width: 40%;
  }
}