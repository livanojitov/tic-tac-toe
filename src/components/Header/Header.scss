.header {
  width: 95%;
  margin: 0 auto;
}

header {
  display: inline-block;
  width: 30%;
  vertical-align: top;
}

.flags img {
  cursor: pointer;
  margin: 0 2%;
}

@media screen and (max-width: 480px) {
  header {
    width: 40%;
  }
}
