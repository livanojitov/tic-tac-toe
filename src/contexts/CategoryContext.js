import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";

export const CategoryContext = createContext();

const CategoryContextProvider = ({ children }) => {
  let { data: categories, error } = useFetchData("categories");

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  return (
    <CategoryContext.Provider value={{ categories }}>
      {children}
    </CategoryContext.Provider>
  );
};

export default CategoryContextProvider;
