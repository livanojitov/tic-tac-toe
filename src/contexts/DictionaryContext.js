import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";
import useLanguage from "../hooks/useLanguage";
import * as constants from "../Constants";

export const DictionaryContext = createContext();

const DictionaryContextProvider = ({ children }) => {
  let { data: dictionary, error } = useFetchData("dictionary");

  const [language, changeLanguageHandler] = useLanguage(constants.ENGLISH);

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  dictionary = dictionary[language];

  return (
    <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
