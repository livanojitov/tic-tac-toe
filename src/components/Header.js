import React, { useContext} from 'react';
import { LanguageContext } from '../contexts/LanguageContext';

const Header = () => { 
  const { setLanguage } = useContext(LanguageContext);
  return (
    <header>
      <div className="flags">
        <img src={'./images/flags/usa.jpg'}    onClick={() => { setLanguage(0); }} alt="Usa"/>
        <img src={'./images/flags/france.jpg'} onClick={() => { setLanguage(1); }}  alt="France"/>
        <img src={'./images/flags/spain.jpg'}  onClick={() => { setLanguage(2); }} alt="Spain"/>      
      </div>
    </header>
  )
}

export default Header;