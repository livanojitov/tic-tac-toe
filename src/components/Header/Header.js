import React, { useContext } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import * as constants from "../../Constants";
import "./Header.scss";

const { ENGLISH, FRENCH, SPANISH } = constants;

const Header = () => {
  const { dictionary, changeLanguageHandler } = useContext(DictionaryContext);

  if (dictionary) {
    const url =
      "https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images";
    const english = dictionary["ENGLISH_HINT"];
    const french = dictionary["FRENCH_HINT"];
    const spanish = dictionary["SPANISH_HINT"];

    return (
      <header>
        <div className="flags" data-testid="flags">
          <img
            data-testid="english"
            src={`${url}%2Fflags%2Fusa.jpg?alt=media`}
            onClick={() => {
              changeLanguageHandler(ENGLISH);
            }}
            alt={english}
            title={english}
          />
          <img
            data-testid="french"
            src={`${url}%2Fflags%2Ffrance.jpg?alt=media`}
            onClick={() => {
              changeLanguageHandler(FRENCH);
            }}
            alt={french}
            title={french}
          />
          <img
            data-testid="spanish"
            src={`${url}%2Fflags%2Fspain.jpg?alt=media`}
            onClick={() => {
              changeLanguageHandler(SPANISH);
            }}
            alt={spanish}
            title={spanish}
          />
        </div>
      </header>
    );
  } else {
    return null;
  }
};

export default Header;
