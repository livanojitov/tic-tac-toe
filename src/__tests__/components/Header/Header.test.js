import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import "@testing-library/jest-dom/extend-expect";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import Header from "../../../components/Header/Header";
import { ENGLISH, FRENCH, SPANISH } from "../../../Constants";

describe("Header", () => {
  let dictionary;
  let changeLanguageHandler;

  beforeEach(() => {
    dictionary = {
      ENGLISH_HINT: "English",
      FRENCH_HINT: "French",
      SPANISH_HINT: "Spanish",
    };
    changeLanguageHandler = jest.fn();
  });

  const renderHeader = () => {
    return render(
      <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
        <Header />
      </DictionaryContext.Provider>
    );
  };

  it("renders the Header component correctly", () => {
    const { asFragment } = renderHeader();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should apply the images's wrapper class", () => {
    renderHeader();

    expect(screen.getByTestId("flags")).toHaveClass("flags");
  });

  it("should not display the flag's images if dictionary hasn't been fetched yet", () => {
    dictionary = "";

    renderHeader();

    expect(screen.queryByTestId("flags")).toBeNull();
  });

  it("should display the flag's images once the dictionary has been fetched", () => {
    renderHeader();

    const flags = screen.getByTestId("flags");
    expect(flags).toContainElement(screen.getByTestId("english"));
    expect(flags).toContainElement(screen.getByTestId("french"));
    expect(flags).toContainElement(screen.getByTestId("spanish"));
  });

  it("should apply english image element alt attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("english")).toHaveAttribute(
      "alt",
      dictionary.ENGLISH_HINT
    );
  });

  it("should apply french image element alt attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("french")).toHaveAttribute(
      "alt",
      dictionary.FRENCH_HINT
    );
  });

  it("should apply spanish image element alt attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("spanish")).toHaveAttribute(
      "alt",
      dictionary.SPANISH_HINT
    );
  });

  it("should apply english image element title attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("english")).toHaveAttribute(
      "title",
      dictionary.ENGLISH_HINT
    );
  });

  it("should apply french image element title attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("french")).toHaveAttribute(
      "title",
      dictionary.FRENCH_HINT
    );
  });

  it("should apply spanish image element title attribute and value", () => {
    renderHeader();

    expect(screen.getByTestId("spanish")).toHaveAttribute(
      "title",
      dictionary.SPANISH_HINT
    );
  });

  it("should change the language to english", () => {
    renderHeader();

    userEvent.click(screen.getByTestId("english"));

    expect(changeLanguageHandler).toHaveBeenCalledTimes(1);
    expect(changeLanguageHandler).toHaveBeenCalledWith(ENGLISH);
  });

  it("should change the language to french", () => {
    renderHeader();

    userEvent.click(screen.getByTestId("french"));

    expect(changeLanguageHandler).toHaveBeenCalledTimes(1);
    expect(changeLanguageHandler).toHaveBeenCalledWith(FRENCH);
  });

  it("should change the language to spanish", () => {
    renderHeader();

    userEvent.click(screen.getByTestId("spanish"));

    expect(changeLanguageHandler).toHaveBeenCalledTimes(1);
    expect(changeLanguageHandler).toHaveBeenCalledWith(SPANISH);
  });
});
