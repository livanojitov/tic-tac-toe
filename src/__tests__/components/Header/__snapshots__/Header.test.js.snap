// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the Header component correctly 1`] = `
<DocumentFragment>
  <header>
    <div
      class="flags"
      data-testid="flags"
    >
      <img
        alt="English"
        data-testid="english"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fflags%2Fusa.jpg?alt=media"
        title="English"
      />
      <img
        alt="French"
        data-testid="french"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fflags%2Ffrance.jpg?alt=media"
        title="French"
      />
      <img
        alt="Spanish"
        data-testid="spanish"
        src="https://firebasestorage.googleapis.com/v0/b/puzzle-ebd10.appspot.com/o/images%2Fflags%2Fspain.jpg?alt=media"
        title="Spanish"
      />
    </div>
  </header>
</DocumentFragment>
`;
